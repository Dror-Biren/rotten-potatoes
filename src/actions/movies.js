//import uuid from 'uuid';
import database from '../firebase/firebase';
import { uploadPoster } from "../firebase/posters";


// ADD_MOVIE
export const addMovie = (movie) => ({
   type: 'ADD_MOVIE',
   movie
});

function addMovieToDatabase(movie, dispatch) {
   return database.ref(`movies`).push(movie)
      .then((ref) =>
         dispatch(addMovie({
            id: ref.key,
            ...movie
         }))
      );
}

export const startAddMovie = (movieData = {}) => {
   return (dispatch, getState) => {
      const {
         title = '',
         description = '',
         rating = 0,
         ratingsAmount = 0,
         releaseDate = 0,
         genres = [],
         poster,
         trailerUrl = ''
      } = movieData;

      const creator = {
         uid: getState().user.uid,
         time: new Date().valueOf()
      }

      const movie = { title, description, rating, ratingsAmount, releaseDate, genres, trailerUrl, creator, raters: {} }

      function afterUploadPoster(newUrl) {
         movie.posterUrl = newUrl;
         return addMovieToDatabase(movie, dispatch);
      }

      if (poster)
         return uploadPoster(poster)
            .then(afterUploadPoster)

      return addMovieToDatabase(movie, dispatch);
   }
}




function editMovieInDatabase(updates, id, dispatch, newRater) {
   let firebaseUpdates = updates;
   if (newRater) {
      const newRaterPath = `raters/${newRater.key}`;
      firebaseUpdates[newRaterPath] = newRater.data;
   }

   return database.ref(`movies/${id}`).update(firebaseUpdates).then(() => {
      dispatch(editMovie(id, updates, newRater));
   });
}

// EDIT_MOVIE
export const editMovie = (id, updates, newRater) => ({
   type: 'EDIT_MOVIE',
   id,
   updates,
   newRater
});

export const startEditMovie = (id, updates) => {
   return (dispatch, getState) => {
      const { poster, newRatingVote, comment } = updates;
      delete updates.poster;
      delete updates.newRatingVote;
      delete updates.comment;

      const newRater = newRatingVote && {
         key: getState().user.uid,
         data: {
            rating: newRatingVote,
            comment,
            time: new Date().valueOf()
         }
      }

      if (poster)
         return uploadPoster(poster)
            .then(afterUploadPoster)

      return editMovieInDatabase(updates, id, dispatch, newRater);

      function afterUploadPoster(newUrl) {
         updates.posterUrl = newUrl;
         return editMovieInDatabase(updates, id, dispatch, newRater);
      }
   };
};



// UPDATE_MOVIE_RATING
export const updateMovieRating = (movie, newRatingVote, comment) => {
   let { id, rating, ratingsAmount } = movie;
   const newRatingSum = ratingsAmount * rating + newRatingVote;
   ratingsAmount++;
   rating = newRatingSum / ratingsAmount;
   return startEditMovie(id, { rating, ratingsAmount, newRatingVote, comment })
}



// REMOVE_MOVIE
export const removeMovie = ({ id } = {}) => ({
   type: 'REMOVE_MOVIE',
   id
});

export const startRemoveMovie = ({ id } = {}) => {
   return (dispatch, getState) => {
      return database.ref(`movies/${id}`).remove().then(() => {
         dispatch(removeMovie({ id }));
      });
   };
};





// SET_MOVIES
export const setMovies = (movies) => ({
   type: 'SET_MOVIES',
   movies
});


