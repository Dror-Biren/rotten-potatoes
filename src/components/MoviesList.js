import React from 'react';
import { connect } from 'react-redux';
import MovieListItem from './MovieListItem';
import selectMovies from '../selectors/movies';

export const MoviesList = (props) => {

  const noMovies = (
    <div className="list-item list-item--message">
      <span>No movie matches your search settings</span>
    </div>
  );

  const Movies = props.movies.map((movie) => 
      <MovieListItem key={movie.id} {...movie} />
  );
  
  return (
  <div className="content-container">
    <div className="list-body">
      { props.movies.length === 0 ? noMovies : Movies}
    </div>
  </div>
)};

const mapStateToProps = (state) => ({
    movies: selectMovies(state.movies, state.filters)
})

export default connect(mapStateToProps)(MoviesList);
