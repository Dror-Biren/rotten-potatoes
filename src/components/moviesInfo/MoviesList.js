import React from 'react';
import { connect } from 'react-redux';

import MovieListItem from './MovieListItem';
import selectMovies from '../../selectors/movies';

export const MoviesList = (props) => {
   const Movies = props.movies.map(movie =>
      <MovieListItem
         key={movie.id}
         {...movie}
      />
   );

   return (
      <div className="page-margin">
         <div className="list-body">
            {Movies}
         </div>
      </div>
   )
};

const mapStateToProps = (state) => ({
   movies: selectMovies(state.movies, state.filters)
})

export default connect(mapStateToProps)(MoviesList);
