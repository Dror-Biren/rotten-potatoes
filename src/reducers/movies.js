// Expenses Reducer

const moviesReducerDefaultState = [];

export default (state = moviesReducerDefaultState, action) => {
   switch (action.type) {
      case 'ADD_MOVIE':
         return [
            ...state,
            action.movie
         ];

      case 'REMOVE_MOVIE':
         const isDiffrentMovie = ({ id }) => id !== action.id;
         return state.filter(isDiffrentMovie);

      case 'EDIT_MOVIE':
         const getEditedMovie = (movie) => {
            if (action.newRater) {
               const { key, data } = action.newRater;
               if (!movie.raters)
                  movie.raters = {};
               movie.raters[key] = data;
            }
            return {
               ...movie,
               ...action.updates
            };
         }

         return state.map((movie) => {
            if (movie.id === action.id)
               return getEditedMovie(movie);
            return movie;
         });     

      case 'SET_MOVIES':
         return action.movies;

      default:
         return state;
   }
};
