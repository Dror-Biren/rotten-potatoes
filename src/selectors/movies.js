import moment from 'moment'

function hasValueInCommon(strings1, strings2) {
   const getStringStart = str => str.substring(0, 4);
   const compareLexicographicOrder = (str1, str2) => (str1 > str2 ? 1 : -1);

   const sortStrings1 = strings1.map(getStringStart).sort(compareLexicographicOrder);
   const sortStrings2 = strings2.map(getStringStart).sort(compareLexicographicOrder);

   let i1 = 0, i2 = 0;
   while (i1 < sortStrings1.length && i2 < sortStrings2.length) {
      if (sortStrings1[i1] === sortStrings2[i2])
         return true;
      if (sortStrings1[i1] < sortStrings2[i2])
         i1++;
      else
         i2++;
   }
   return false;
}

// Get visible movies
export default (movies, { text, sortBy, yearsRange, genres }) => {
   const [startYear, endYear] = yearsRange;

   function isMovieMatch(movie) {
      //console.log(movie.raters);
      const releaseYear = moment(movie.releaseDate).year();

      const startYearMatch = startYear <= releaseYear;
      const endYearMatch = endYear >= releaseYear;
      const textMatch = movie.title.toLowerCase().includes(text.toLowerCase());
      const genresMatch = hasValueInCommon(genres, movie.genres);

      return startYearMatch && endYearMatch && textMatch && genresMatch;
   }

   function sortMovies(movie1, movie2) {
      switch (sortBy) {
         case 'viewes':
            return movie2.ratingsAmount - movie1.ratingsAmount;
         case 'date':
            return movie2.releaseDate - movie1.releaseDate;
         case 'rating':
            return movie2.rating - movie1.rating;
      }
   }

   return movies.filter(isMovieMatch).sort(sortMovies);
};
