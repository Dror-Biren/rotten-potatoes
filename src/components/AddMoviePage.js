import React from 'react';
import { connect } from 'react-redux';
import MovieForm from './MovieForm';
import { startAddMovie } from '../actions/movies';

export class AddMoviePage extends React.Component {
   returnToMainPage = () => {
      this.props.history.push('/');
   };

   onSubmit = (movie) => {
      this.props.startAddMovie(movie);
      this.props.history.push('/');
   };

   render() {
      return (
         <div>
            <div className="sub-header">
               
                  <h2 className="sub-header__title">
                     Add a new movie to the site
                  </h2>
               
               <img
                  src="/images/goBack5.png"
                  className="button--go-back"
                  onClick={this.returnToMainPage}>
               </img>
            </div>
            <div className="page-margin">
               <MovieForm
                  onSubmit={this.onSubmit}
               />
            </div>
         </div>
      );
   }
}

const mapDispatchToProps = (dispatch) => ({
   startAddMovie: (movie) => dispatch(startAddMovie(movie))
});

export default connect(undefined, mapDispatchToProps)(AddMoviePage);
