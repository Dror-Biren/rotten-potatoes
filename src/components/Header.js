import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogin } from '../actions/auth';
import { startLogout } from '../actions/auth';
import { history } from '../routers/AppRouter';

export const Header = ({ isAuthenticated, startLogin, startLogout }) => {
   const logOutButton = (
      <button className="button" onClick={startLogout}>
         Log-out
      </button>
   );

   const logInButton = (
      <button className="button" onClick={startLogin}>
         Login with Google
      </button>
   );

   return (
      <header className="header">
         <img
            className="header__background"
            src="/images/potatoes.webp"
         />
         <div className="content-container">
            <div className="header__content">

               <Link className="header__title" to="/dashboard">
                  <h1>
                     Rotten Potatoes
                  </h1>
               </Link>

               { isAuthenticated ? logOutButton : logInButton }
            </div>
         </div>
      </header>
   );
}

const mapStateToProps = (state) => ({
   isAuthenticated: !!state.auth.uid
 });

const mapDispatchToProps = (dispatch) => ({
   startLogin: () => dispatch(startLogin()),
   startLogout: () => {
      dispatch(startLogout()).then();
      const returnToLoginPage = () => history.push('/');
      setTimeout(returnToLoginPage, 700);
   }
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
