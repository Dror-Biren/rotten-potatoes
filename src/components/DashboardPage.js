import React from 'react';

import MoviesList from './MoviesList';
import MovieListFilters from './MovieListFilters';
import MoviesSummary from './MoviesSummary';

const DashboardPage = () => (
  <div>
    <MoviesSummary />
    <MovieListFilters />
    <MoviesList />
  </div>
);

export default DashboardPage;
