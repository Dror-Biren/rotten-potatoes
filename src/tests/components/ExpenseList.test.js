import React from 'react';
import { shallow } from 'enzyme';
import { MoviesList } from '../../components/MoviesList';
import movies from '../fixtures/movies';

test('should render MoviesList with movies', () => {
  const wrapper = shallow(<MoviesList movies={movies} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render MoviesList with empty message', () => {
  const wrapper = shallow(<MoviesList movies={[]} />);
  expect(wrapper).toMatchSnapshot();
});
