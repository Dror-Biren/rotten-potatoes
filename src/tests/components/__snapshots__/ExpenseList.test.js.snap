// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MoviesList with empty message 1`] = `
<div
  className="page-margin"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Movie
    </div>
    <div
      className="show-for-desktop"
    >
      rating
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No movies
      </span>
    </div>
  </div>
</div>
`;

exports[`should render MoviesList with movies 1`] = `
<div
  className="page-margin"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Movie
    </div>
    <div
      className="show-for-desktop"
    >
      rating
    </div>
  </div>
  <div
    className="list-body"
  >
    <MovieListItem
      rating={195}
      releaseDate={0}
      title="Gum"
      id="1"
      description=""
    />
    <MovieListItem
      rating={109500}
      releaseDate={-345600000}
      title="Rent"
      id="2"
      description=""
    />
    <MovieListItem
      rating={4500}
      releaseDate={345600000}
      title="Credit Card"
      id="3"
      description=""
    />
  </div>
</div>
`;
